src/index.tsx

```diff
@@ -0,0 +1,131 @@
+import invariant from 'invariant';
+import React, { useCallback, useContext, useState } from 'react';
+import { ReactRelayContext, commitMutation } from 'react-relay';
+import { MutationConfig, OperationBase } from 'relay-runtime';
+
+import { Omit, WithOptionalFields } from './typeHelpers';
+
+export type MutationState<T extends OperationBase> = {
+  loading: boolean;
+  data: T['response'] | null;
+  error?: Error | null;
+};
+
+export type Mutate<T extends OperationBase> = (
+  config?: Partial<MutationConfig<T>>,
+) => void;
+
+type MutationConfigWithOptionalVariables<
+  T extends OperationBase
+> = WithOptionalFields<MutationConfig<T>, 'variables'>;
+
+export function useMutation<T extends OperationBase>(
+  userConfig: MutationConfigWithOptionalVariables<T>,
+): [Mutate<T>, MutationState<T>] {
+  const [state, setState] = useState<MutationState<T>>({
+    loading: false,
+    data: null,
+    error: null,
+  });
+
+  const { environment } = useContext(ReactRelayContext);
+  const {
+    configs,
+    mutation,
+    variables,
+    uploadables,
+    onCompleted,
+    onError,
+    optimisticUpdater,
+    optimisticResponse,
+    updater,
+  } = userConfig;
+
+  const mutate = useCallback(
+    (config?: Partial<Omit<MutationConfig<T>, 'mutation'>>) => {
+      const mergedConfig = {
+        configs,
+        mutation,
+        variables,
+        uploadables,
+        onCompleted,
+        onError,
+        optimisticUpdater,
+        optimisticResponse,
+        updater,
+        ...config,
+      };
+
+      invariant(mergedConfig.variables, 'you must specify variables');
+
+      setState({
+        loading: true,
+        data: null,
+        error: null,
+      });
+
+      return new Promise((resolve, reject) => {
+        commitMutation(environment, {
+          ...mergedConfig,
+          variables: mergedConfig.variables!,
+          onCompleted: (response, error) => {
+            invariant(!error, 'mutation unexpectedly completed with error');
+
+            setState({
+              loading: false,
+              data: response,
+              error: null,
+            });
+
+            if (mergedConfig.onCompleted) {
+              mergedConfig.onCompleted(response, undefined);
+            } else {
+              resolve(response);
```

Comment:
i think we can just always resolve... no need for `else` here. we don't want to always reject because unhandled rejections = uncaught exceptions

Reply:
I guess. unresolved promises are as bad as unhandled rejections though, no? don't they lead to leak?

Reply:
no, they're fine... if they have no callbacks they get GCed normally. the issue w/the uncaught rejection is really the crash handling. otherwise any call to an async fn would introduce problems